core/src/price_estimation/clients/oneinch/api.rs

```diff
@@ -0,0 +1,179 @@
+use super::super::generic_client::{Api, GenericToken};
+use crate::http::{HttpClient, HttpFactory, HttpLabel};
+use anyhow::{Context, Result};
+use ethcontract::Address;
+use futures::future::{BoxFuture, FutureExt as _};
+use serde::Deserialize;
+use std::collections::HashMap;
+use url::Url;
+
+#[derive(Clone, Debug, Deserialize, PartialEq)]
+pub struct Token {
+    pub name: String,
+    pub symbol: String,
+    pub address: Option<Address>,
+    pub decimals: u8,
+}
+impl GenericToken for Token {
+    fn symbol(&self) -> &str {
+        &self.symbol
+    }
+}
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct TradedAmounts {
+    to_token_amount: String,
+    from_token_amount: String,
+}
+
+#[derive(Debug)]
+pub struct OneinchHttpApi {
+    api_url: Url,
+    client: HttpClient,
+}
+impl OneinchHttpApi {
+    pub fn with_url(http_factory: &HttpFactory, api_url: &str) -> Result<Self> {
+        let client = http_factory
+            .create()
+            .context("failed to initialize HTTP client")?;
+        let api_url = api_url
+            .parse()
+            .with_context(|| format!("failed to parse url {}", api_url))?;
+        Ok(OneinchHttpApi { api_url, client })
+    }
+}
+
+pub const DEFAULT_API_URL: &str = "https://api.1inch.exchange/v1.1/";
+
+/// 1inch API version v1.1
+/// https://1inch.exchange/#/api
+impl Api for OneinchHttpApi {
+    type Token = Token;
+
+    fn bind(http_factory: &HttpFactory) -> Result<Self> {
+        Self::with_url(http_factory, DEFAULT_API_URL)
+    }
+
+    fn get_token_list<'a>(&'a self) -> BoxFuture<'a, Result<Vec<Token>>> {
+        async move {
+            let url = self.api_url.join("tokens").unwrap();
+            let token_mapping: HashMap<String, Token> = self
+                .client
+                .get_json_async(url.to_string(), HttpLabel::Oneinch)
+                .await
+                .context("failed to parse token list json from 1inch response")?;
+            Ok(token_mapping
+                .into_iter()
+                .map(|(_token_symbol, token_data)| token_data)
+                .collect())
+        }
+        .boxed()
+    }
+
+    fn get_price<'a>(&'a self, from: &Token, to: &Token) -> BoxFuture<'a, Result<f64>> {
+        let mut url = self.api_url.join("quote").unwrap();
+        // 1inch requires the user to specify the amount traded in atoms.
+        // We compute the price when selling one full token to avoid unavoidable rounding
+        // artifacts when selling exactly one token atom.
+        let mut one_token_from = String::from("1");
+        one_token_from.push_str(&String::from("0").repeat(from.decimals as usize));
+        {
+            // This is in its own block because we are supposed to drop `q`.
+            let mut q = url.query_pairs_mut();
+            q.append_pair("fromTokenSymbol", &from.symbol);
+            q.append_pair("toTokenSymbol", &to.symbol);
+            q.append_pair("amount", &one_token_from);
```

Comment:
nit: `query_pairs_mut` returns a `Serializer<UrlQuery>` which supports the builder pattern, so you may even be able to:
```suggestion
            url.query_pairs_mut()
                .append_pair("fromTokenSymbol", &from.symbol)
                .append_pair("toTokenSymbol", &to.symbol)
                .append_pair("amount", &one_token_from);
```

This might make it so you don't need the block since you never create a binding to the intermediate binding.

Reply:
Nice! I also changed it in dex.ag.